### PR TITLE
Docker support

### PR DESCRIPTION
**Pull request by bazsi**
This makes it pretty easy to compile rpms and debs locally using a docker supported build infrastructure.

The docker images are not yet available in dockerhub, so you'll need to build them locally:
```
$ dbld/rules images
```

To use, make sure you have the contents of the "release" branch:
```
$ git merge release
```

Then simply build the debs:
```
$ dbld/rules deb-jessie
```

More help is displayed when using "dbld/rules" without arguments.

### CODE REVIEW

**Comment by kira-syslogng** [comment]
Build SUCCESS, the tests were executed on test branch: master and test suite: functions

**Comment by bazsi** [comment]
@lbudai can we merge this before the release? It does not touch code and makes it a lot easier to compile syslog-ng by end users.

It does need improvements (like different distros) but I think would help on the binary problem.

**Comment by mitzkia** [comment]
Hi, 

I have tested the scripts. 
The docker image building works correctly (images are built).
When I wanted to compile syslog-ng inside docker image (balabit/syslog-ng-zesty), I used the following commands:
```
$ dbld/rules shell-zesty
$ ./autogen.sh
$ cd /build
$ /source/configure
$ make
```
But without `make install` syslog-ng won't be installed. In this case, we have the following problem:
 - docker image started with a local (non root) user, this user can not write into default install path: `/usr/local`
 - console log
```
 /bin/mkdir -p '/usr/local/lib'
 /bin/bash ./libtool   --mode=install /usr/bin/install -c   lib/libsyslog-ng.la '/usr/local/lib'
libtool: install: /usr/bin/install -c lib/.libs/libsyslog-ng-3.9.so.0.0.0 /usr/local/lib/libsyslog-ng-3.9.so.0.0.0
/usr/bin/install: cannot create regular file '/usr/local/lib/libsyslog-ng-3.9.so.0.0.0': Permission denied
Makefile:6303: recipe for target 'install-libLTLIBRARIES' failed
make[3]: *** [install-libLTLIBRARIES] Error 1
Makefile:16149: recipe for target 'install-am' failed
make[2]: *** [install-am] Error 2
Makefile:14984: recipe for target 'install-recursive' failed
make[1]: *** [install-recursive] Error 1
Makefile:16143: recipe for target 'install' failed
make: *** [install] Error 2
```

**Comment by bazsi** [comment]
hi,

thanks for the feedback. the primary use-case for the images is to easily
build rpm/deb packages. install is a valid use-case too though, but I want
to leave the "configure" script in the hands of the user, which means that
he/she has to specify --prefix to a proper location.

I've just pushed a patch that does that, what do you think?


-- 
Bazsi

**Comment by kira-syslogng** [comment]
Build SUCCESS, the tests were executed on test branch: master and test suite: functions

**Comment by mitzkia** [comment]
What do you think about the continuous testing of the docker images? I now that this question is not a scope of this PR.
I mean the images can be work in one day and can be broken in another day, for example, because of new dependencies. I would be good if we can setup the docker autobuild for the images.
[https://docs.docker.com/docker-hub/builds/](url)

**Comment by kira-syslogng** [comment]
Build SUCCESS, the tests were executed on test branch: master and test suite: functions

**Comment by bazsi** [comment]
Can you help setting up the automatic builds? I mean it would be great if
they didnt need building locally.

**Comment by kira-syslogng** [comment]
Build SUCCESS, the tests were executed on test branch: master and test suite: functions

**Comment by bazsi** [comment]
@mitzkia can you supply your approve? this could go in then.

**Comment by kira-syslogng** [comment]
Build SUCCESS, the tests were executed on test branch: master and test suite: functions

**Comment by mitzkia** [comment]
Yes, of course, I can help setting up automatic builds.
After thinking about it again we have following option:
1, docker auto build service can help us to create daily syslog-ng build environments (balabit/syslog-ng-jessie, balabit/syslog-ng-zesty, balabit/syslog-ng-centos7)
2, Kira can help us to approve that within our daily build environments we can still build syslog-ng (or generate native packages).
I also have other ideas, but it is not this PR's scope.

I don't know if I can approve this PR alone. I would wait for others to comment it also.
Anyway, I have got some other issues during the build of native packages. From this, I will have some inline notes, but I need some time to investigate them.

File: dbld/rpm [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+	echo "  1) you don't do this from a release tarball"
+	echo "  2) you haven't merged the 'release' branch"
+	echo "  3) you didn't create rpm packaging manually"
+	echo ""
+	echo "Please make sure you have obtained a source distribution with a spec file and rerun this command"
+	exit 1
+fi
+VERSION=`cat VERSION`
+
+cd /build
```
Comment (mitzkia) [inline dbld/rpm]: This line was needed for me here to start the `rpmbuild` command:
`cp syslog-ng-${VERSION}.tar.gz syslog-ng_${VERSION}.tar.gz`

File: dbld/images/centos7/dev-dependencies.txt [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+# other tools
+gdb
+git
+lsof
+strace
+vim
+# debian packaging tools
+docbook-xsl
+python-nose
+python-ply
```
Comment (mitzkia) [inline dbld/images/centos7/dev-dependencies.txt]: This package was needed for to start the `rpmbuild`:
gcc-c++

**Comment by bazsi** [comment]
Hi,

Maybe I didnt push the "release" branch, which contains these fixes, as
they are in the spec file.

File: dbld/images/centos7/dev-dependencies.txt (hunk 2 above)
Comment (bazsi) [inline dbld/images/centos7/dev-dependencies.txt]: how did you get the rpm packaging stuff? this seems to be the "old" one on the master branch.

you'd need to merge in the release branch, as both debian/rpm packaging is on those branches. It's a hassle, which we should discuss with @mochrul and @lbudai.

while staying on the "docker-support" branch, just do "git merge release" and you'd have all the files. Later you can reset them away.

Comment (bazsi) [inline dbld/images/centos7/dev-dependencies.txt]: btw, I've checked, my changes of the release branch is already on github.

Comment (mitzkia) [inline dbld/images/centos7/dev-dependencies.txt]: Sorry, it was my fault. After started it from scratch it works.

File: dbld/rpm (hunk 1 above)
Comment (mitzkia) [inline dbld/rpm]: Sorry, it was my fault. After started it from scratch it works.

File: dbld/deb [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+VERSION=`cat VERSION`
+
+cd /build
+
+tar xvf syslog-ng-${VERSION}.tar.gz
+cp syslog-ng-${VERSION}.tar.gz syslog-ng_${VERSION}.orig.tar.gz
+cd syslog-ng-${VERSION}
+dpkg-buildpackage -rfakeroot
+
+echo "Your debs are in /build, also available on the host in \$(top_srcdir)/dlbd/build"
```
Comment (mitzkia) [inline dbld/deb]: Instead of dlbd -> dbld.

File: dbld/rpm [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+rpmbuild -ta syslog-ng-${VERSION}.tar.gz
+
+echo "Your rpms are in /build, also available on the host in \$(top_srcdir)/dlbd/build"
```
Comment (mitzkia) [inline dbld/rpm]: Instead of dlbd -> dbld.

File: dbld/tarball [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+
+cd /source
+VERSION=`cat VERSION`
+./autogen.sh
+
+cd /build
+/source/configure
+make dist
+
+echo "Your tarball is in /build, also available on the host in \$(top_srcdir)/dlbd/build"
```
Comment (mitzkia) [inline dbld/tarball]: Instead of dlbd -> dbld.

File: dbld/rpm [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+	echo "  2) you haven't merged the 'release' branch"
+	echo "  3) you didn't create rpm packaging manually"
+	echo ""
+	echo "Please make sure you have obtained a source distribution with a spec file and rerun this command"
+	exit 1
+fi
+VERSION=`cat VERSION`
+
+cd /build
+rpmbuild -ta syslog-ng-${VERSION}.tar.gz
```
Comment (mitzkia) [inline dbld/rpm]: I guess we forgot to copy the generated rpm-s into the /build directory. 
The rpmbuild command ended with this lines: 
```
Checking for unpackaged file(s): /usr/lib/rpm/check-files /home/micek/rpmbuild/BUILDROOT/syslog-ng-3.9.1-20.el7.centos.x86_64
Wrote: /home/micek/rpmbuild/SRPMS/syslog-ng-3.9.1-20.el7.centos.src.rpm
Wrote: /home/micek/rpmbuild/RPMS/x86_64/syslog-ng-3.9.1-20.el7.centos.x86_64.rpm
Wrote: /home/micek/rpmbuild/RPMS/x86_64/syslog-ng-libdbi-3.9.1-20.el7.centos.x86_64.rpm
Wrote: /home/micek/rpmbuild/RPMS/x86_64/syslog-ng-mongodb-3.9.1-20.el7.centos.x86_64.rpm
Wrote: /home/micek/rpmbuild/RPMS/x86_64/syslog-ng-smtp-3.9.1-20.el7.centos.x86_64.rpm
Wrote: /home/micek/rpmbuild/RPMS/x86_64/syslog-ng-java-3.9.1-20.el7.centos.x86_64.rpm
Wrote: /home/micek/rpmbuild/RPMS/x86_64/syslog-ng-geoip-3.9.1-20.el7.centos.x86_64.rpm
Wrote: /home/micek/rpmbuild/RPMS/x86_64/syslog-ng-redis-3.9.1-20.el7.centos.x86_64.rpm
Wrote: /home/micek/rpmbuild/RPMS/x86_64/syslog-ng-riemann-3.9.1-20.el7.centos.x86_64.rpm
Wrote: /home/micek/rpmbuild/RPMS/x86_64/syslog-ng-http-3.9.1-20.el7.centos.x86_64.rpm
Wrote: /home/micek/rpmbuild/RPMS/x86_64/syslog-ng-python-3.9.1-20.el7.centos.x86_64.rpm
Wrote: /home/micek/rpmbuild/RPMS/x86_64/syslog-ng-devel-3.9.1-20.el7.centos.x86_64.rpm
Wrote: /home/micek/rpmbuild/RPMS/x86_64/syslog-ng-debuginfo-3.9.1-20.el7.centos.x86_64.rpm
Executing(%clean): /bin/sh -e /var/tmp/rpm-tmp.2ICEHz
+ umask 022
+ cd /home/micek/rpmbuild/BUILD
+ cd syslog-ng-3.9.1
+ /usr/bin/rm -rf /home/micek/rpmbuild/BUILDROOT/syslog-ng-3.9.1-20.el7.centos.x86_64
+ exit 0
Your rpms are in /build, also available on the host in $(top_srcdir)/dlbd/build
```
The rpm packages are under: `/home/micek/rpmbuild/RPMS/x86_64/`.

File: dbld/shell [hunk 7]
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
+
+echo -e "\n\n\nWelcome to the syslog-ng developer's shell"
+
+echo -e "\tYou can find the source in /souce, you can build the source using:"
```
Comment (presidento) [inline dbld/shell]: typo: `/souce` (missing `r`)

**Comment by lbudai** [comment]
@bazsi: my plan is to release 3.10 on next week, so yes, it is possible to merge this PR.

**Comment by kira-syslogng** [comment]
Build SUCCESS, the tests were executed on test branch: master and test suite: functions

**Comment by kira-syslogng** [comment]
Build SUCCESS, the tests were executed on test branch: master and test suite: functions

**Comment by kira-syslogng** [comment]
Build SUCCESS, the tests were executed on test branch: master and test suite: functions

**Comment by bazsi** [comment]
the clang build in travis was successful in one case and failed next time just because it failed to resolve java dependencies due to a network connectivity error.

With the approval of @lbudai IRL, I am merging this, even before 3.10, as it is pretty much independent from the code.

**Pull request merged**